### PR TITLE
Add example yaml files for deployment and testing

### PR DESCRIPTION
**Pull request by xing-yang**
This PR adds example yaml files for deployment and testing of the snapshot feature.

Images will be updated when they are released.

### CODE REVIEW

File: examples/pvc.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+metadata:
+  name: hpvc
+spec:
+  storageClassName: csi-hostpath-sc
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 1Gi
+
```
Comment (wackxu) [inline examples/pvc.yaml]: remove blank line

File: examples/restore.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  dataSource:
+    name: new-snapshot-demo 
+    kind: VolumeSnapshot
+    apiGroup: snapshot.storage.k8s.io
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 1Gi
+
```
Comment (wackxu) [inline examples/restore.yaml]: ditto

File: examples/storageclass.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+metadata:
```
Comment (wackxu) [inline examples/storageclass.yaml]: storageclass?

File: deploy/kubernetes/setup-csi-snapshotter.yaml [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+ 
+---
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: external-snapshotter-runner
+rules:
+  - apiGroups: [""]
+    resources: ["persistentvolumes"]
+    verbs: ["get", "list", "watch", "create", "delete"]
```
[the remark this answers is not in the document]
Comment (jingxu97) [reply deploy/kubernetes/setup-csi-snapshotter.yaml]: but the error is for list PVCs, you don't need create/delete ?

File: deploy/kubernetes/setup-csi-snapshotter.yaml [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    verbs: ["get", "list"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotclasses"]
+    verbs: ["get", "list", "watch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotcontents"]
+    verbs: ["create", "get", "list", "watch", "update", "delete"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshots"]
+    verbs: ["create", "get", "list", "watch", "update", "delete"]
```
Comment (jingxu97) [inline deploy/kubernetes/setup-csi-snapshotter.yaml]: no need to have create?

File: deploy/kubernetes/setup-csi-snapshotter.yaml [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshots/status"]
```
Comment (jingxu97) [inline deploy/kubernetes/setup-csi-snapshotter.yaml]: with the above permission on volumesnapshots, I don't think you need to put /status here?

File: deploy/kubernetes/setup-csi-snapshotter.yaml [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+            - name: ADDRESS
+              value: /csi/csi.sock
+          imagePullPolicy: "IfNotPresent"
+          volumeMounts:
+            - name: socket-dir
+              mountPath: /csi
+        - name: csi-snapshotter
+          image: xyang105/csi:csi-snapshotter
+          args:
+            - "--snapshotter=csi-hostpath"
```
Comment (jingxu97) [inline deploy/kubernetes/setup-csi-snapshotter.yaml]: I think we don't need to specify this driver name any more.

File: deploy/kubernetes/setup-csi-snapshotter.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (xing-yang) [reply deploy/kubernetes/setup-csi-snapshotter.yaml]: Looks like I didn't read the comments carefully:).  I'll verify.

**Comment by wackxu** [comment]
@xing-yang @jingxu97 I saw there are in-tree clusterRole for external-provisioner and external-attacher, We should also add this for snapshotter, if it is ok, I can add this.

File: deploy/kubernetes/setup-csi-snapshotter.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (xing-yang) [reply deploy/kubernetes/setup-csi-snapshotter.yaml]: Verified that I need to keep create/delete too:

I0915 03:26:41.931543       1 controller.go:1113] provision "default/hpvc" class "csi-hostpath-sc": failed to save persistentvolume "pvc-1e1481d2-b897-11e8-944c-000c2967769a": persistentvolumes is forbidden: User "system:serviceaccount:default:csi-snapshotter" cannot create resource "persistentvolumes" in API group "" at the cluster scope

File: deploy/kubernetes/setup-csi-snapshotter.yaml (hunk 7 above)
Comment (xing-yang) [reply deploy/kubernetes/setup-csi-snapshotter.yaml]: We use this in SetDefaultSnapshotClass. Do you mean this is no longer needed if we have admission control?  If that is what you meant, we still need it now until it is set automatically.

To be consistent with the provisioner, I think we should set it too.

File: deploy/kubernetes/setup-csi-snapshotter.yaml (hunk 5 above)
Comment (xing-yang) [reply deploy/kubernetes/setup-csi-snapshotter.yaml]: Done

File: deploy/kubernetes/setup-csi-snapshotter.yaml (hunk 6 above)
Comment (xing-yang) [reply deploy/kubernetes/setup-csi-snapshotter.yaml]: Done

**Comment by xing-yang** [comment]
@wackxu Can you add a link to those here?

**Comment by wackxu** [comment]
@xing-yang When I add in-tree test, I found this, I have add the code and test code.

https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L460-L469

https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L504-L513

File: deploy/kubernetes/setup-csi-snapshotter.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (wackxu) [reply deploy/kubernetes/setup-csi-snapshotter.yaml]: @xing-yang Yes, you are right,  we deploy external-provisioner together with external-snapshotter, so we need this.

**Comment by xing-yang** [comment]
I don't know if they would let us add this for an out-of-tree CRD.  The one they added is CRD used by kubelet so it is different.

Jing, what do you think?

File: deploy/kubernetes/setup-csi-snapshotter.yaml (hunk 7 above)
Comment (jingxu97) [reply deploy/kubernetes/setup-csi-snapshotter.yaml]: https://github.com/kubernetes-csi/external-snapshotter/blob/master/cmd/csi-snapshotter/main.go#L118
snapshotter driver name is returned by this function. User does not need to set it manually.

Comment (xing-yang) [reply deploy/kubernetes/setup-csi-snapshotter.yaml]: Sure.  Will remove it.

Comment (xing-yang) [reply deploy/kubernetes/setup-csi-snapshotter.yaml]: Fixed

**Comment by jingxu97** [comment]
For ClusterRole question, since snapshots are CRD, currently we don't need to have in-tree clusterRole for them.

**Pull request merged**